libcxx/test/libcxx/algorithms/alg.sorting/assert.sort.invalid_comparator/assert.sort.invalid_comparator.pass.cpp

```diff
@@ -0,0 +1,195 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// REQUIRES: has-unix-headers
+// UNSUPPORTED: c++03, c++11, c++14, c++17
+// UNSUPPORTED: !libcpp-hardening-mode=debug
+// XFAIL: libcpp-hardening-mode=debug && availability-verbose_abort-missing
+
+// This test uses a specific combination of an invalid comparator and sequence of values to
+// ensure that our sorting functions do not go out-of-bounds and satisfy strict weak ordering in that case.
+// Instead, we should fail loud with an assertion. The specific issue we're looking for here is when the comparator
+// does not satisfy the strict weak ordering:
+//
+//    Irreflexivity: comp(a, a) is false
+//    Antisymmetry: comp(a, b) implies that !comp(b, a)
+//    Transitivity: comp(a, b), comp(b, c) imply comp(a, c)
+//    Transitivity of equivalence: !comp(a, b), !comp(b, a), !comp(b, c), !comp(c, b) imply !comp(a, c), !comp(c, a)
+//
+// If this is not satisfied, we have seen issues in the past where the std::sort implementation
+// would proceed to do OOB reads. (rdar://106897934).
+// Other algorithms like std::stable_sort, std::sort_heap do not go out of bounds but can produce
+// incorrect results, we also want to assert on that.
+// Sometimes std::sort does not go out of bounds as well, for example, right now if transitivity
+// of equivalence is not met, std::sort can only produce incorrect result but would not fail.
+
+#include <algorithm>
+#include <cassert>
+#include <cstddef>
+#include <limits>
+#include <map>
+#include <memory>
+#include <ranges>
+#include <random>
+#include <set>
+#include <string>
+#include <vector>
+
+#include "bad_comparator_values.h"
+#include "check_assertion.h"
+#include "invalid_comparator_utilities.h"
+
+void check_oob_sort_read() {
+  SortingFixture fixture(SORT_DATA);
+
+  // Check the classic sorting algorithms
+  {
+    std::vector<std::size_t*> copy = fixture.create_elements();
+    TEST_LIBCPP_ASSERT_FAILURE(
+        std::sort(copy.begin(), copy.end(), fixture.checked_predicate()),
+        "Would read out of bounds, does your comparator satisfy the strict-weak ordering requirement?");
```

Comment:
We should figure out why this doesn't fail earlier with the strict weak ordering checks, but this can be done in a follow-up.

Reply:
Thanks for filing the issue! Wrote a comment -- briefly, this seems to be expected given our current state, but our current state is probably not ideal. What I'm trying to say, this is a little unexpected but nothing seems horribly broken.